schematics/playground-module/add-to-modules.ts

```diff
@@ -0,0 +1,246 @@
+ */
+
+import * as ts from 'typescript';
+import { dirname, Path } from '@angular-devkit/core';
+import { DirEntry, SchematicContext, SchematicsException, Tree } from '@angular-devkit/schematics';
+import { getSourceFile } from '@angular/cdk/schematics';
+import { addImportToModule, addProviderToModule, getDecoratorMetadata } from '@schematics/angular/utility/ast-utils';
+import {
+  addRoute,
+  generateComponentRoute,
+  importPath,
+  getClassWithDecorator,
+  multilineArrayLiteral,
+  removeExtension,
+  getServicesFromDir,
+  applyInsertChange,
+  hasRoutingModuleInDir,
+  addDeclaration,
+  getComponentsFromDir,
+  getDirectivesFromDir,
+  getFeatureModuleFromDir,
+  getRoutingModuleFromDir,
+  generateLazyModulePath,
+  routePredicatesFromPath,
+  applyReplaceChange,
+  findRoutingModule,
+  findFeatureModule,
+  getPlaygroundRootDir,
+  findRoutesArray,
+  addMissingChildRoutes,
+  isInPlaygroundRoot,
+  getRouteLazyModule,
+  addObjectProperty,
+  findRouteWithPath,
+  generatePathRoute,
+  findRoute,
+  componentRoutePredicate,
+  lazyModulePredicate,
+} from '../utils';
+
+export function addToModules(tree: Tree, context: SchematicContext): Tree {
+  processDirs(tree, context, [getPlaygroundRootDir(tree)]);
+  return tree;
+}
+
+function processDirs(tree: Tree, context: SchematicContext, dirs: DirEntry[]): void {
+  for (const dir of dirs) {
+    processDir(tree, context, dir);
+
+    if (dir.subdirs.length > 0) {
+      processDirs(tree, context, dir.subdirs.map(d => dir.dir(d)));
+    }
+  }
+}
+
+function processDir(tree: Tree, context: SchematicContext, dir: DirEntry): void {
+  getComponentsFromDir(dir).forEach(path => processComponent(tree, context, path));
+  getDirectivesFromDir(dir).forEach(path => processDirectives(tree, path));
+  getServicesFromDir(dir).forEach(path => processService(tree, path));
+
+  const modulePath = getFeatureModuleFromDir(dir);
+  if (modulePath) {
+    processFeatureModule(tree, context, modulePath);
+  }
+
+  const routingModulePath = getRoutingModuleFromDir(dir);
+  if (routingModulePath) {
+    processRoutingModule(tree, context, routingModulePath);
+  }
+}
+
+function processService(tree: Tree, servicePath: Path): void {
+  const modulePath = findFeatureModule(tree, dirname(servicePath));
+  if (modulePath == null) {
+    throw new SchematicsException(`Can't find module for service ${servicePath}.`);
+  }
+  const serviceDeclarations = getClassWithDecorator(tree, servicePath, 'Injectable');
+
+  for (const service of serviceDeclarations) {
+    const serviceClassName = (service.name as ts.Identifier).getText();
+    const importString = importPath(modulePath, servicePath);
+    const source = getSourceFile(tree, servicePath);
+    const changes = addProviderToModule(source, modulePath, serviceClassName, importString);
+
+    applyInsertChange(tree, modulePath, ...changes);
+  }
+}
+
+function processComponent(tree: Tree, context: SchematicContext, componentPath: Path): void {
```

Comment:
All `process*` methods and especially this one are big and hard to understand what they do. Maybe, it would be better to split them somehow?

Reply:
Also, feel free to add some comments 😺